### ISSUE TITLE
Semi-Random Events

### ISSUE THREAD

**Issue by Wrongtown** [daid/EmptyEpsilon#155]
I saw issue #154 (AI Chatter) and thought it similar enough to something I was working on that I would mention it and see what people think. I haven't had much time lately to work on it so probably best I check that people don't think it's a terrible idea before I make time to pick it up again :smile: 

## Core Concept
Other things are happening in the universe, and players will often have an opportunity to respond to these events in a way that feels meaningful.

I have started writing a script snippet/module intended to produce semi-randomly selected events at semi-random intervals using math.randomseed. Basically if it has been more than X time since one of these random events was triggered, generate a random number from a range and run the scenario corresponding to that number. Some events would be more likely than others by virtue of having a greater number of values in the range that lead to them. 

I'm thinking probably best not to raise another until existing side-quest is resolved if these are used during scenarios that have a "Main Mission" but you might make them more frequent or allow overlap for use in a Sandbox. 

Examples of a few similar scenarios/decision trees and consequences: 
## EVENT TYPE 1 
Players intercept an SOS broadcast from a stranded ship (spawned at semi-random location far enough from player that they can't tell it just pops into existance). I planned to include some technobabble about a "transponder" or something hinting that the SOS is coming from a friendly/enemy as appropriate

### Initial information
1.a1 Ship identifies as Player Faction
1.b1 Ship identifies as Player Faction
1.c1 Ship identifies as Enemy Faction (_"Oh my stars whatever will we our intrepid Captain decide to do about this?"_)

### Examples of what is needed
1.*2i Ship requests a Homing missile so they can retrofit their ship using its guidance module to limp to a station for proper repairs.
1.*2ii Ship requests an energy transfer to <technobabble> their <technobabble>
1.*2iii Ship is too far gone and crew request transfer to player ship and will be "dropped off" when players next dock at specific/appropriate station (Reputation+++ for this one, possibly add tension with ticking timer to get close enough to transfer crew before ship explodes damaging all around it - possibly simulate this by spawning a mine/nuke on the location? Haven't tested if that works) - If ship-to-ship docking possible I would use this but otherwise they just have to get really close.

### Players can respond in several ways:
	a) Say they'll help and follow through
	b) Say they'll help but fail to follow through
	c) Say they won't help
	d) Say they won't help in a rude or aggressive way

### Consequences
1.a3 (Identified as player faction, actually is player faction)
	a) Reputation++ and if there's still a ship it will respond enthusiastically to commands (once situation resolved). May also give some kind of reward (as long as it's not the same thing they were asking for).
	bi) No change to Reputation
	bii) No change to Reputation (After X time ship goes away, maybe someone else rescued them, maybe they died, THESE ARE THE QUESTIONS THAT KEEP A CAPTAIN AWAKE DURING THEIR SLEEP-CYCLE)
	c) Reputation- with disappointed response
	d) Reputation-- with very disappointed response

1.b3 (Identified as player faction, actually is enemy faction)
	ai) Arrive at coordinates to find it's a trap, ship is fine but controlled by enemies and attacks players
	aii) Arrive at coordinates to find it's a trap, ship is fine and attacks players along with additional enemies to jump in
	bi) No change to Reputation (enemies wait paitently to spring trap)
	bii) No change to Reputation (after X time enemies get bored of waiting and start moving to attack player/player faction)
	biii) No change to Reputation (after X time enemies get bored of waiting and disappear)
	c) Reputation- (temporarily!) with disappointed response while "stranded" ship starts moving towards players (still not obviously enemy, revealed either by attacking or Deep Scan from Science as it gets closer).
	d) Enemy commander loses cool and shouts insults, giving up on stealth revealing trap and starts moving to attack player ship.

1c.3 (Identified as enemy faction)
	ai) Enemy ship thanks players and skulks off (will not attack player ship unless fired upon, maybe even will help against enemies that aren't of same faction? Was thinking basically would set this ship to player Faction. See footnote.)
	aii) Enemy ship thanks players and gives them some kind of reward (Reputation+++/energy or weapons as long as it isn't what they needed from player ship - I think there's an opportunity here for some great rewards in the future)
	aiii) It's a trap! (See 1.b3)
	bi) No change to Reputation
	bii) No change to Reputation (After X time ship goes away)
	c) Sad acknowledgement that this is to be expected, no change to Reputation (Some may prefer to have this set to a reputation drop especially if you're more into the roleplay aspect and your crew are expected to be Paragons of Virtue)
	d) Angry acknowledgement that this is to be expected but that the enemy commander thinks you're scum for kicking them while they're down (may not be appropriate for some enemy factions?).

## EVENT TYPE 2
Players intercept a signal that the computer indicated was a message, but when you open the frequency you hear only the haunting sounds of a pod of Space Whales... (No decision required here folks, just a bit of filler and the timer resets for another event to be triggered in X.)

_**Footnote:** I kind of like the idea that this sort of thing could even lead to a whole fleet of "enemy" ships turning up lead by this commander to help players in certain types of scenarios. This could be a single use "Favour" triggered by Relay or more of an unexpected "Han Solo swooping in to help with the Death Star" sort of thing._

**Comment by Fouindor** [comment]
Dabbled with the idea when I started writing scripts for EE, I had started to cobble together some kind of "framework" for event generation and handling for a random "border patrol" mission, never got around finishing it, only the reward attribution code is fully functional as of now :
http://img15.hostingpics.net/pics/864081rec.png
http://img15.hostingpics.net/pics/593457rec1.png

However, it's a given for me that random events should be optional, as they may interfere with scripted missions. If you're interested with adding events, I can finish and release it.

**Comment by nallath** [comment]
Some scenarios already have "random" npc's flying about. You could simply chance decide when it's underway if some mishap occurs with them. That way you never have the issue of ships just popping into existance.

**Comment by Wrongtown** [comment]
@Fouindor hey that's cool, is the pop-up shown in your screenshots something that's available to all stations? I totally agree that it should be optional, I was thinking this would be a resource people could add to their scenarios to flesh them out a little. Alternatively it could be a standalone script that you enable an option for at server launch if it's popular enough.

@nallath this is true, but my intent behind summoning the entities into existence "off-camera" is that way it can work agnostic of whether there is a handy NPC to grab.

**Comment by nallath** [comment]
That's true, but i'd argue that there should have been npc's in the first place. Random events are nicer when they are a logic conclusion of chance npc meetings.

**Comment by Wrongtown** [comment]
@nallath how would you see this working? I am thinking it would require that scripts include a setNPC("allow")/setNPC("block") or similar tag so you're not going to have some Random Event interfere with a vital element of the main scenario plot.

Off the top of my head that would add twice as many gates to pass through for an event to start. For example the current dependencies are:
1. Have we reached the eventTriggerTimer threshold?
YES
2. Is there already an unresolved randomEvent?
NO
3. Did we randomly generate a trigger number?
YES
>Launch the event that corresponds to the trigger number.

We would need to also check:
4. Were any NPC entities flagged in this scenario?
YES
5. Are any of the NPC entities flagged in this scenario still "alive"?
YES
6. Do any of the living NPC entities belong to the faction required for this event to make sense?
YES
7. Are any of the NPC entities which meet 4, 5 & 6 of the correct type (station/ship) required for this event to make sense?
YES
>Launch the event that corresponds to the trigger number.

While I do like the idea of always having random NPCs going about their business in scripts, I'm not sure it's worth demanding that every script creator includes them in order for this to work without breaking the "plot". I did start thinking along the lines of front-loading the whole process by generating our own NPC entities at the beginning so we eliminate step 4, but step 5 is still a concern.

Of course, there could be an elegant solution that I'm totally missing here, but unless there is I think utilizing existing NPC entities falls into the "Nice To Have" bucket rather than "Must Have" for this feature. Doesn't the "off-camera" spawn provide the illusion of a fuller universe at a much lower cost?

**Comment by ajford** [comment]
I agree with @Wrongtown. Spawning off-camera seems like a relatively easy way to make the universe seem a lot more full without a lot of effort. 

If you wrote this up as a module that other script-builders could use, this would be AWESOME! If you do this, I recommend you put it together as a github repo, then people could submit PRs with new scenarios.

**Comment by nallath** [comment]
There is actually quite a bit of logic required to handle off screen spawning. Sure for one player ship it's doable, but for multiple it gets quite a bit more tricky.

You can also get very weird situations where a ship is near a nebula (where it can't see through). It just passed point A (which it can see). It flies to point B (where it can't see point A). Then it flies back to point A, only to suddenly find a stranded ship that wasn't there before. How did it get there then?

**Comment by daid** [comment]
Untested:
```
function findRandomPointOutOfPlayerSight()
    local ships = {}
    for n=1,32 do
        local ship = getPlayerShip(n)
        if ship ~= nil then
            table.insert(ships, ship)
        end
    end
    while true do
        local ship = ships[irandom(1, #ships)]
        local a = random(0, 360)
        local length = random(30000, 35000)
        local x, y = ship:getPosition()
        x = x + math.cos(a / 180 * math.pi) * length
        y = y + math.sin(a / 180 * math.pi) * length
        local allow = true
        for _, ship in pairs(ships) do
            if distance(ship, x, y) < 30000 then
                allow = false
            end
        end
        if allow then
            return x, y
        end
    done
end
```

Should find a random position outside of the radar range of any player.

**Comment by kwadroke** [comment]
That's if you don't adjust the Radar Range, correct?

**Comment by daid** [comment]
True, but we can add a simple function to request this range.

**Comment by MyrddinE** [comment]
Note: This function could cause the enemy ship to spring into being on the HUD of the Relay player, if the point is within 5km of an allied ship.

**Comment by Wrongtown** [comment]
Good thinking, though I'd say that's just when the allied ship shared the information ;)

Honestly though, I'd be pretty surprised if an average player (rather than the subset of players here who are constantly peeking under the hood) on Relay managed to notice the exact moment when a ship popped up into existence with such clarity of situational awareness that it messed with their suspension of disbelief. 

Also any ship with a jump drive would pretty much do exactly that, as long as the flavour-text doesn't carry on about "We've been here for weeks!" then even a ship with a busted drive just becomes a ship with a Very Recently busted drive ;)

Sorry I haven't been contributing more, things are a bit hectic right now! I haven't even revisited those tutorial edits, but will do.

**Comment by daid** [comment]
Or when they jumped in from hyperspace?

The beacon of light scenario spawns an ambush right in your face, none of the players I tried it with found that confusing. It does that right after you make a jump, so that helps I guess.

**Comment by nallath** [comment]
I've noticed this behaviour happening with various GTA games (and actually used it to spawn vehicles when I need one). Simply looking down & up again would quite often conjure cars out of thin air.

**Comment by Wrongtown** [comment]
I don't know what the equivalent of looking down then up again would be here but I guess if you _want_ I could add it to the list and figure out how later? :stuck_out_tongue_winking_eye:
